src/codegen/spirv.zig

```diff
@@ -1351,80 +1207,36 @@ const NavGen = struct {
         };
 
         // Kernel only supports unsigned ints.
-        if (self.getTarget().os.tag == .vulkan) {
-            return self.spv.intType(signedness, backing_bits);
+        if (self.spv.hasFeature(.kernel)) {
+            return self.spv.intType(.unsigned, backing_bits);
         }
 
-        return self.spv.intType(.unsigned, backing_bits);
+        return self.spv.intType(signedness, backing_bits);
     }
 
     fn arrayType(self: *NavGen, len: u32, child_ty: IdRef) !IdRef {
-        // TODO: Cache??
-        const len_id = try self.constInt(Type.u32, len, .direct);
-        const result_id = self.spv.allocId();
-
-        try self.spv.sections.types_globals_constants.emit(self.spv.gpa, .OpTypeArray, .{
-            .id_result = result_id,
-            .element_type = child_ty,
-            .length = len_id,
-        });
-        return result_id;
-    }
-
-    fn ptrType(self: *NavGen, child_ty: Type, storage_class: StorageClass) !IdRef {
-        return try self.ptrType2(child_ty, storage_class, .indirect);
+        const len_id = try self.constInt(Type.u32, len);
+        return self.spv.arrayType(len_id, child_ty);
     }
 
-    fn ptrType2(self: *NavGen, child_ty: Type, storage_class: StorageClass, child_repr: Repr) !IdRef {
-        const key = .{ child_ty.toIntern(), storage_class, child_repr };
-        const entry = try self.ptr_types.getOrPut(self.gpa, key);
-        if (entry.found_existing) {
-            const fwd_id = entry.value_ptr.ty_id;
-            if (!entry.value_ptr.fwd_emitted) {
-                try self.spv.sections.types_globals_constants.emit(self.spv.gpa, .OpTypeForwardPointer, .{
-                    .pointer_type = fwd_id,
-                    .storage_class = storage_class,
-                });
-                entry.value_ptr.fwd_emitted = true;
-            }
-            return fwd_id;
-        }
-
-        const result_id = self.spv.allocId();
-        entry.value_ptr.* = .{
-            .ty_id = result_id,
-            .fwd_emitted = false,
-        };
-
+    fn ptrType(self: *NavGen, child_ty: Type, storage_class: StorageClass, child_repr: Repr) !IdRef {
         const child_ty_id = try self.resolveType(child_ty, child_repr);
-
-        try self.spv.sections.types_globals_constants.emit(self.spv.gpa, .OpTypePointer, .{
-            .id_result = result_id,
-            .storage_class = storage_class,
-            .type = child_ty_id,
-        });
-
+        const result_id = try self.spv.ptrType(child_ty_id, storage_class);
+        if (storage_class == .Uniform or storage_class == .PushConstant) try self.spv.decorate(child_ty_id, .Block);
+        try self.spv.decorate(result_id, .{ .ArrayStride = .{ .array_stride = @intCast(child_ty.abiSize(self.pt.zcu)) } });
         return result_id;
     }
```

Comment:
This is not the right way to handle things because it doesn't correctly handle recursive pointers. The call to `self.resolveType()` would enter an infinite loop, thats why the `ptr_types` map is handled especially. For example, this kernel fails:
```
const Node = struct { node: *Node };

export fn oef() callconv(.spirv_kernel) void {
    var a: Node = undefined;
    _ = &a;
}
```
```
~/programming/zig/build/spirv-zig/bin/zig build-obj ./test.zig  -fno-compiler-rt -target spirv64-ope
ncl-gnu -mcpu generic+int64+int16+int8+float64+float16+generic_pointer+v1_6+kernel -fno-llvm -freference-trace
fish: Job 1, '~/programming/zig/build/spirv-z…' terminated by signal SIGSEGV (Address boundary error)
```

Reply:
In the future we might be able to remove this ugly piece of code by relying on `SPV_KHR_untyped_pointers`. Some projects are already adding support for this, so hopefully we will be able to use this in the not too far away future.